IBM/couchbackup
# Experimental attachments option

## Checklist

- [x] Added tests for code changes _or_ test/build only changes
- [x] Updated the change log file (`CHANGES.md`) _or_ test/build only changes
- [x] Completed the PR template below:

## Description

Add experimental attachments option, supersedes #733 

## Approach



* Add `quiet` as `boolean` type to `app.js` validation because it was missing previously
* Pass `options` object through to `Backup` and `Restore` classes so that the `attachments` option can be processed.
* Add attachments handling
    * Refactor stream pipeline setup to allow attachments mapper to be inserted between source and destination
    * Add attachments metadata to backup file
    * Add errors for restoring with/without attachments option depending on presence of backup file attachments flag
    * Add `attachments: true` parameter to `_all_docs` and `_bulk_docs` requests if attachments option enabled
    * Add attachment mappers to encode/decode Buffer/Base64 (although inline attachments are base64 the SDK automatically decodes them to buffers for convenience - so they need re-encoding to store as base64 string in the backup file)
* Add `attachments` option
    * emit warning when experimental option is used
    * add option to CLI scripts
    * add option to CLI parser
    * add `attachments: false` option to defaults
* Update documentation, with a lot of emphasis that the option is experimental and unsupported

## Schema & API Changes

* Adds experimental `attachments` option that can be supplied via CLI, env var or options object.
* Adds `attachments` metadata to backup file "header".

## Security and Privacy

- "No change"

## Testing

- Added new tests:
    - attachments mapping tests for buffer/base64 conversions
    - added backup and restore mapping cases for attachments option
    - added shallow backup test with attachments option
    - add `attachments` option checks for CLI parameter and env vars
    - `test/app.js` add `quiet` option checks that were missing
    - `should restore and backup attachment` end to end test that restores a backup including an attachment, then backs up the restored database and compares the backup file to the original test fixture.
- Modified existing tests to:
    -  pass empty options object to `Backup`/`Restore` ctors.
    - add `attachments` metadata to test fixtures

## Monitoring and Logging

- Added new log lines for `couchbackup:mappers` for attachment processing
